### PR TITLE
csrf protection in job_agent

### PR DESCRIPTION
**Pull request by richo-anyscale**
r? @alanwguo 

## Why are these changes needed?

Prevent confused deputy attacks on the job_agent

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

(Working on the checks!)

### CODE REVIEW

File: python/ray/dashboard/optional_utils.py [hunk 1]
```diff
@@ -261,6 +262,21 @@ def _update_cache(task):
...
+def deny_browser_requests() -> Callable:
+    """Reject any requests that appear to be made by a browser"""
+    def decorator_factory(f: Callable) -> Callable:
+        @functools.wraps(f)
+        async def decorator(self, req: Request):
+            if browser_detection.is_browser_request(req):
+                return Response(
+                        text="Method not allowed",
+                        status=aiohttp.web.HTTPNotAllowed.status_code,
+                        )
```
Comment (alanwguo) [inline python/ray/dashboard/optional_utils.py:274]: Should we tell people that browser requests are not supported? Or do we purposely not expose this detail?

Comment (richo-anyscale) [reply python/ray/dashboard/optional_utils.py:274]: I don't think there's meaningful value to hiding it although you are so far from a supported path if you wind up there. I can adjust the message.

**Comment by richo-anyscale** [comment]
@alanwguo could you have another look? The implementation wound up changing quite a bit.

**Comment by richo-anyscale** [comment]
bumps, @aslonnie @alanwguo @thomasdesr do you think someone could have a look for me?

File: python/ray/tests/test_dashboard.py [hunk 2]
```diff
@@ -4,11 +4,11 @@
 import sys
 import time
 
+import ray
```
Comment (aslonnie) [inline python/ray/tests/test_dashboard.py:7]: probably revert the change of this file?

File: python/ray/dashboard/optional_utils.py [hunk 3]
```diff
@@ -260,6 +260,28 @@ def _update_cache(task):
         return _wrapper
 
 
+def is_browser_request(req: Request) -> bool:
```
Comment (aslonnie) [inline python/ray/dashboard/optional_utils.py:263]: maybe add a doc string?

**Review by aslonnie — APPROVED** [review]
I do not own these parts.. though the change looks good to my eyes.

**Review by alanwguo — APPROVED** [review]
